src/Router.ts

```diff
@@ -0,0 +1,186 @@
+import { after } from 'dojo-core/aspect';
+import Evented from 'dojo-core/Evented';
+import { Handle, EventObject } from 'dojo-core/interfaces';
+import Promise from 'dojo-core/Promise';
+import CancelNavigationError from './errors/CancelNavigationError';
+import has from './has';
+import HashSource from './HashSource';
+import HtmlHistorySource from './HtmlHistorySource';
+import PathRule from './PathRule';
+import RouteManager from './RouteManager';
+import { CancelableNavigationArgs, MatchableRoute, NavigationArgs, Route, RouteHandlers, RouterArgs, RouterSource } from './routing';
+
+const DefaultSource = has('history') ? HtmlHistorySource : HashSource;
+let defaultSource: RouterSource = null;
+
+if (has('host-browser')) {
+	defaultSource = new (<any> DefaultSource)();
+}
+
+// TODO Replace with `lang.assign` when it's available.
+function mixin(target: {}, ...sources: {}[]): {} {
+	return sources.reduce(function (object: {}, source: {}): {} {
+		Object.keys(source).forEach(function (key: string): void {
+			(<any> object)[key] = (<any> source)[key];
+		});
+
+		return object;
+	}, target);
+}
+
+export default class Router extends RouteManager {
+	protected _evented: Evented;
+	protected _isRun: boolean;
+
+	canceled: NavigationArgs;
+	current: NavigationArgs;
+	error: (error: Error) => void;
```

Comment:
If this were to be removed in a favor of an "error" event (which IMO it should), then the promise returned by `Router#go` should both emit this error *and* reject, correct?